src/ssh/sftp_dir_iterator.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2022 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/format.h>
+#include <multipass/ssh/sftp_dir_iterator.h>
+#include <multipass/ssh/sftp_utils.h>
+
+#include <stdexcept>
+
+namespace multipass
+{
+void SFTPDirIterator::push_dir(const std::string& path)
+{
+    auto dir = mp_sftp_opendir(sftp, path.c_str());
+    if (!dir)
+        throw std::runtime_error{
+            fmt::format("[sftp] cannot open remote directory '{}': {}", path, ssh_get_error(sftp->session))};
+
+    dirs.push(std::move(dir));
+}
+
+SFTPDirIterator::SFTPDirIterator(sftp_session sftp, const fs::path& path) : sftp(sftp)
+{
+    push_dir(path.u8string());
+    next();
+}
+
+bool SFTPDirIterator::hasNext() const
+{
+    return (bool)previous_attr;
+}
+
+SFTPAttributesUPtr SFTPDirIterator::next()
+{
+    if (dirs.empty())
+        return std::move(previous_attr);
+
+    auto dir = dirs.top().get();
+    while (auto attr = mp_sftp_readdir(sftp, dir))
+    {
+        if (strcmp(attr->name, ".") == 0 || strcmp(attr->name, "..") == 0)
+            continue;
+
+        auto path = fmt::format("{}/{}", dir->name, attr->name);
```

Comment:
It shouldn't be tied to the host OS. `SFTPDirIterator` iterates over the directories inside the VM, not locally, and the VM always runs Linux :smile: 
This was actually a big hassle to handle, as I initially used the `std::filesystem::path` default behavior everywhere, which always uses OS specific separators. On Windows, of course, this will create paths with '\', but then working with those paths inside the VM through SFTP wouldn't make much sense, since '\' will not be used as a path separator. That's why I had to go through the entire code again and find all paths that are going to be used inside the VM and make sure they always use '/'. Honestly sometimes I regret using `std::filesystem::path` instead of plain old `std::string` with '/', just because of this pesky little '\'...

Reply:
OK, don't mind me then. Ideally the SFTP code shouldn't assume the remote OS, but I guess that's fine in here. We can generalize if we ever have the need.